File: apps/desktop/src/lib/project-settings.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';

const ID_PATTERN = /^[0-9a-f]{64}$/;
const POSITIVE_DECIMAL_PATTERN = /^[1-9][0-9]*$/;
const LOWER_HEX_PATTERN = /^(?:[0-9a-f]{2})+$/;
const MAX_PERSISTED_INTEGER = 9_223_372_036_854_775_807n;
const MAX_IGNORE_PATTERNS = 256;
const MAX_IGNORE_PATTERN_BYTES = 1_024;
const MAX_COMMANDS = 256;
const MAX_EXECUTABLE_BYTES = 4 * 1_024;
const MAX_ARGUMENTS = 256;
const MAX_ARGUMENT_BYTES = 4 * 1_024;
const MAX_TOTAL_ARGUMENT_BYTES = 64 * 1_024;
const MAX_WORKING_DIRECTORY_HEX_LENGTH = 2 * 131_072;

export type DiscoveredCommandKindV1 = 'test' | 'build' | 'lint' | 'format';

export interface ProjectIgnoreSettingsV1 {
  configurationPresent: boolean;
  patterns: string[];
}

export interface DiscoveredCommandV1 {
  arguments: string[];
  commandId: string;
  evidenceCount: number;
  executable: string;
  kind: DiscoveredCommandKindV1;
  selected: boolean;
  workingDirectoryHex: string | null;
}

export type ProjectCommandConfirmationV1 =
  | { status: 'notConfirmed' }
  | { confirmedAtUnixMillis: string; revision: string; status: 'current' | 'stale' };

export type ProjectCommandSettingsV1 =
  | { status: 'noPublishedIndex' }
  | {
      catalogId: string;
      commands: DiscoveredCommandV1[];
      confirmation: ProjectCommandConfirmationV1;
      status: 'available';
    };

export interface ActiveProjectSettingsV1 {
  commands: ProjectCommandSettingsV1;
  ignore: ProjectIgnoreSettingsV1;
}

export type ProjectSettingsResultV1 =
  { status: 'noProject' } | { settings: ActiveProjectSettingsV1; status: 'available' };

export interface ProjectSettingsResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: ProjectSettingsResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryProjectSettings(
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<ProjectSettingsResponseV1> {
  const payload = await invokeCommand('query_project_settings', {
    request: { protocolVersion: CURRENT_PROTOCOL_VERSION },
  });
  return parseProjectSettingsResponseV1(payload);
}

export async function confirmProjectCommandAllowlist(
  expectedCatalogId: string,
  expectedAllowlistRevision: string | null,
  commandIds: string[],
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<ProjectSettingsResponseV1> {
  assertId(expectedCatalogId, 'Command catalog ID');
  if (expectedAllowlistRevision !== null) {
    assertPositiveDecimal(expectedAllowlistRevision, 'Command allowlist revision');
  }
  if (
    commandIds.length === 0 ||
    commandIds.length > MAX_COMMANDS ||
    new Set(commandIds).size !== commandIds.length
  ) {
    throw new Error('Command selection is empty, duplicated, or outside the V1 bound.');
  }
  commandIds.forEach((commandId) => assertId(commandId, 'Command ID'));

  const payload = await invokeCommand('confirm_project_command_allowlist', {
    request: {
      commandIds: [...commandIds].sort(),
      expectedAllowlistRevision,
      expectedCatalogId,
      protocolVersion: CURRENT_PROTOCOL_VERSION,
    },
  });
  return parseProjectSettingsResponseV1(payload);
}

export function parseProjectSettingsResponseV1(payload: unknown): ProjectSettingsResponseV1 {
  if (
    !isRecord(payload) ||
    !hasExactKeys(payload, ['protocolVersion', 'result']) ||
    payload.protocolVersion !== CURRENT_PROTOCOL_VERSION
  ) {
    throw new Error('Project Settings response does not match the V1 schema.');
  }
  return {
    protocolVersion: payload.protocolVersion,
    result: parseResult(payload.result),
  };
}

function parseResult(value: unknown): ProjectSettingsResultV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Project Settings response contains an invalid result.');
  }
  if (value.status === 'noProject' && hasExactKeys(value, ['status'])) {
    return { status: 'noProject' };
  }
  if (value.status !== 'available' || !hasExactKeys(value, ['settings', 'status'])) {
    throw new Error('Project Settings response contains an invalid result.');
  }
  return { settings: parseActiveSettings(value.settings), status: 'available' };
}

function parseActiveSettings(value: unknown): ActiveProjectSettingsV1 {
  if (!isRecord(value) || !hasExactKeys(value, ['commands', 'ignore'])) {
    throw new Error('Project Settings response contains an invalid active-project snapshot.');
  }
  return { commands: parseCommands(value.commands), ignore: parseIgnore(value.ignore) };
}

function parseIgnore(value: unknown): ProjectIgnoreSettingsV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['configurationPresent', 'patterns']) ||
    typeof value.configurationPresent !== 'boolean' ||
    !Array.isArray(value.patterns) ||
    value.patterns.length > MAX_IGNORE_PATTERNS
  ) {
    throw new Error('Project Settings response contains invalid ignore settings.');
  }
  const patterns = value.patterns.map((pattern) => {
    if (
      typeof pattern !== 'string' ||
      pattern.length === 0 ||
      utf8Length(pattern) > MAX_IGNORE_PATTERN_BYTES ||
      hasControlCharacter(pattern)
    ) {
      throw new Error('Project Settings response contains an invalid ignore pattern.');
    }
    return pattern;
  });
  if (!value.configurationPresent && patterns.length !== 0) {
    throw new Error('Missing project configuration cannot own ignore patterns.');
  }
  return { configurationPresent: value.configurationPresent, patterns };
}

function parseCommands(value: unknown): ProjectCommandSettingsV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Project Settings response contains invalid command settings.');
  }
  if (value.status === 'noPublishedIndex' && hasExactKeys(value, ['status'])) {
    return { status: 'noPublishedIndex' };
  }
  if (
    value.status !== 'available' ||
    !hasExactKeys(value, ['catalogId', 'commands', 'confirmation', 'status']) ||
    typeof value.catalogId !== 'string' ||
    !Array.isArray(value.commands) ||
    value.commands.length > MAX_COMMANDS
  ) {
    throw new Error('Project Settings response contains invalid command settings.');
  }
  assertId(value.catalogId, 'Command catalog ID');
  const commands = value.commands.map(parseCommand);
  if (new Set(commands.map((command) => command.commandId)).size !== commands.length) {
    throw new Error('Project Settings response repeats a command ID.');
  }
  const confirmation = parseConfirmation(value.confirmation);
  const selectedCount = commands.filter((command) => command.selected).length;
  if (
    (confirmation.status === 'current' && selectedCount === 0) ||
    (confirmation.status !== 'current' && selectedCount !== 0)
  ) {
    throw new Error('Project Settings response contains an inconsistent command selection.');
  }
  return {
    catalogId: value.catalogId,
    commands,
    confirmation,
    status: 'available',
  };
}

function parseCommand(value: unknown): DiscoveredCommandV1 {
  const keys = [
    'arguments',
    'commandId',
    'evidenceCount',
    'executable',
    'kind',
    'selected',
    'workingDirectoryHex',
  ];
  if (
    !isRecord(value) ||
    !hasExactKeys(value, keys) ||
    typeof value.commandId !== 'string' ||
    typeof value.executable !== 'string' ||
    utf8Length(value.executable) === 0 ||
    utf8Length(value.executable) > MAX_EXECUTABLE_BYTES ||
    hasControlCharacter(value.executable) ||
    !isCommandKind(value.kind) ||
    typeof value.selected !== 'boolean' ||
    !Number.isInteger(value.evidenceCount) ||
    (value.evidenceCount as number) < 1 ||
    (value.evidenceCount as number) > 16 ||
    !Array.isArray(value.arguments) ||
    value.arguments.length > MAX_ARGUMENTS ||
    !isWorkingDirectoryHex(value.workingDirectoryHex)
  ) {
    throw new Error('Project Settings response contains an invalid discovered command.');
  }
  assertId(value.commandId, 'Command ID');
  let totalArgumentBytes = 0;
  const arguments_ = value.arguments.map((argument) => {
    if (typeof argument !== 'string' || argument.includes('\0')) {
      throw new Error('Project Settings response contains an invalid command argument.');
    }
    const bytes = utf8Length(argument);
    if (bytes > MAX_ARGUMENT_BYTES) {
      throw new Error('Project Settings response contains an oversized command argument.');
    }
    totalArgumentBytes += bytes;
    return argument;
  });
  if (totalArgumentBytes > MAX_TOTAL_ARGUMENT_BYTES) {
    throw new Error('Project Settings response contains an oversized argv.');
  }
  return {
    arguments: arguments_,
    commandId: value.commandId,
    evidenceCount: value.evidenceCount as number,
    executable: value.executable,
    kind: value.kind,
    selected: value.selected,
    workingDirectoryHex: value.workingDirectoryHex,
  };
}

function parseConfirmation(value: unknown): ProjectCommandConfirmationV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Project Settings response contains an invalid command confirmation.');
  }
  if (value.status === 'notConfirmed' && hasExactKeys(value, ['status'])) {
    return { status: 'notConfirmed' };
  }
  if (
    (value.status !== 'current' && value.status !== 'stale') ||
    !hasExactKeys(value, ['confirmedAtUnixMillis', 'revision', 'status']) ||
    typeof value.revision !== 'string' ||
    typeof value.confirmedAtUnixMillis !== 'string'
  ) {
    throw new Error('Project Settings response contains an invalid command confirmation.');
  }
  assertPositiveDecimal(value.revision, 'Command allowlist revision');
  assertNonNegativeDecimal(value.confirmedAtUnixMillis, 'Confirmation timestamp');
  return {
    confirmedAtUnixMillis: value.confirmedAtUnixMillis,
    revision: value.revision,
    status: value.status,
  };
}

function isWorkingDirectoryHex(value: unknown): value is string | null {
  return (
    value === null ||
    (typeof value === 'string' &&
      value.length <= MAX_WORKING_DIRECTORY_HEX_LENGTH &&
      LOWER_HEX_PATTERN.test(value))
  );
}

function isCommandKind(value: unknown): value is DiscoveredCommandKindV1 {
  return value === 'test' || value === 'build' || value === 'lint' || value === 'format';
}

function assertId(value: string, label: string): void {
  if (!ID_PATTERN.test(value)) throw new Error(`${label} is not canonical.`);
}

function assertPositiveDecimal(value: string, label: string): void {
  if (!POSITIVE_DECIMAL_PATTERN.test(value) || BigInt(value) > MAX_PERSISTED_INTEGER) {
    throw new Error(`${label} is not a bounded positive decimal.`);
  }
}

function assertNonNegativeDecimal(value: string, label: string): void {
  if (!/^(?:0|[1-9][0-9]*)$/.test(value) || BigInt(value) > MAX_PERSISTED_INTEGER) {
    throw new Error(`${label} is not a bounded canonical decimal.`);
  }
}

function utf8Length(value: string): number {
  return new TextEncoder().encode(value).length;
}

function hasControlCharacter(value: string): boolean {
  return Array.from(value).some((character) => {
    const codePoint = character.codePointAt(0);
    return (
      codePoint !== undefined && (codePoint <= 0x1f || (codePoint >= 0x7f && codePoint <= 0x9f))
    );
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const keys = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    keys.length === sortedExpected.length &&
    keys.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/project_settings.rs
use crate::ProtocolVersion;
use serde::de::{Error as DeError, SeqAccess, Visitor};
use serde::{Deserialize, Serialize};
use std::fmt;

/// Pathless read request for active-project ignore and safe-command Settings.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryProjectSettingsRequestV1 {
    protocol_version: ProtocolVersion,
}

impl QueryProjectSettingsRequestV1 {
    /// Returns the schema version checked before Core-owned project selection.
    #[must_use]
    pub const fn protocol_version(self) -> ProtocolVersion {
        self.protocol_version
    }
}

/// Exact catalog-bound selection without project, argv, evidence, or timestamp authority.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ConfirmProjectCommandAllowlistRequestV1 {
    protocol_version: ProtocolVersion,
    expected_catalog_id: String,
    expected_allowlist_revision: Option<String>,
    #[serde(deserialize_with = "deserialize_command_ids")]
    command_ids: Vec<String>,
}

fn deserialize_command_ids<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    struct CommandIdsVisitor;

    impl<'de> Visitor<'de> for CommandIdsVisitor {
        type Value = Vec<String>;

        fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
            formatter.write_str("between one and 256 lowercase command IDs")
        }

        fn visit_seq<A>(self, mut sequence: A) -> Result<Self::Value, A::Error>
        where
            A: SeqAccess<'de>,
        {
            if sequence.size_hint().is_some_and(|hint| hint > 256) {
                return Err(A::Error::custom("too many command IDs"));
            }
            let mut values = Vec::with_capacity(sequence.size_hint().unwrap_or(0).min(256));
            while let Some(value) = sequence.next_element::<String>()? {
                if values.len() == 256 {
                    return Err(A::Error::custom("too many command IDs"));
                }
                if value.len() != 64
                    || !value
                        .bytes()
                        .all(|byte| byte.is_ascii_digit() || matches!(byte, b'a'..=b'f'))
                {
                    return Err(A::Error::custom("invalid command ID"));
                }
                values.push(value);
            }
            if values.is_empty() {
                return Err(A::Error::custom("command selection is empty"));
            }
            Ok(values)
        }
    }

    deserializer.deserialize_seq(CommandIdsVisitor)
}

impl ConfirmProjectCommandAllowlistRequestV1 {
    /// Returns the request schema version.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }

    /// Returns the exact current catalog identity shown by the UI.
    #[must_use]
    pub fn expected_catalog_id(&self) -> &str {
        &self.expected_catalog_id
    }

    /// Returns the latest visible append-only confirmation revision, if one exists.
    #[must_use]
    pub fn expected_allowlist_revision(&self) -> Option<&str> {
        self.expected_allowlist_revision.as_deref()
    }

    /// Returns only selected current command IDs; argv remains Core-owned.
    #[must_use]
    pub fn command_ids(&self) -> &[String] {
        &self.command_ids
    }
}

/// Read-only validated `.a3/project.toml` exclusion projection.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectIgnoreSettingsV1 {
    configuration_present: bool,
    patterns: Vec<String>,
}

impl ProjectIgnoreSettingsV1 {
    /// Creates a bounded projection after adapter and Application validation.
    #[must_use]
    pub const fn new(configuration_present: bool, patterns: Vec<String>) -> Self {
        Self {
            configuration_present,
            patterns,
        }
    }
}

/// Closed category of a manifest-evidenced direct-argv command.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum DiscoveredCommandKindV1 {
    /// Project test command.
    Test,
    /// Project build command.
    Build,
    /// Static diagnostic command.
    Lint,
    /// Formatting check command.
    Format,
}

/// One exact bounded direct-argv template from current published evidence.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DiscoveredCommandV1 {
    command_id: String,
    kind: DiscoveredCommandKindV1,
    working_directory_hex: Option<String>,
    executable: String,
    arguments: Vec<String>,
    evidence_count: u16,
    selected: bool,
}

impl DiscoveredCommandV1 {
    /// Creates a safe display projection from the current Core-owned catalog.
    #[allow(clippy::too_many_arguments)]
    #[must_use]
    pub const fn new(
        command_id: String,
        kind: DiscoveredCommandKindV1,
        working_directory_hex: Option<String>,
        executable: String,
        arguments: Vec<String>,
        evidence_count: u16,
        selected: bool,
    ) -> Self {
        Self {
            command_id,
            kind,
            working_directory_hex,
            executable,
            arguments,
            evidence_count,
            selected,
        }
    }
}

/// Durable confirmation state relative to the current evidence-sensitive catalog.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum ProjectCommandConfirmationV1 {
    /// No command subset has ever been confirmed for this worktree.
    NotConfirmed,
    /// Latest confirmation exactly matches current manifest evidence.
    Current {
        /// Monotone private-store CAS revision.
        revision: String,
        /// Core-generated confirmation time.
        #[serde(rename = "confirmedAtUnixMillis")]
        confirmed_at_unix_millis: String,
    },
    /// A prior selection exists but manifest evidence changed.
    Stale {
        /// Monotone private-store CAS revision required to replace it.
        revision: String,
        /// Core-generated time of the now-stale confirmation.
        #[serde(rename = "confirmedAtUnixMillis")]
        confirmed_at_unix_millis: String,
    },
}

/// Command Settings await a publication or present the complete current bounded catalog.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum ProjectCommandSettingsV1 {
    /// The worktree has not published its first deterministic index yet.
    NoPublishedIndex,
    /// A current command catalog and its confirmation state are available.
    Available {
        /// Evidence-sensitive identity of the complete displayed catalog.
        #[serde(rename = "catalogId")]
        catalog_id: String,
        /// At most 256 canonical command templates.
        commands: Vec<DiscoveredCommandV1>,
        /// Latest durable selection relative to this catalog.
        confirmation: ProjectCommandConfirmationV1,
    },
}

/// Complete project-owned Settings projection.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ActiveProjectSettingsV1 {
    ignore: ProjectIgnoreSettingsV1,
    commands: ProjectCommandSettingsV1,
}

impl ActiveProjectSettingsV1 {
    /// Groups the two project settings domains without exposing a project path or identity.
    #[must_use]
    pub const fn new(ignore: ProjectIgnoreSettingsV1, commands: ProjectCommandSettingsV1) -> Self {
        Self { ignore, commands }
    }
}

/// Core-owned project selection result.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum ProjectSettingsResultV1 {
    /// No validated worktree is active.
    NoProject,
    /// Dedicated project Settings were reconstructed safely.
    Available {
        /// Complete bounded settings projection.
        settings: Box<ActiveProjectSettingsV1>,
    },
}

/// Versioned project Settings query or mutation response.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectSettingsResponseV1 {
    protocol_version: ProtocolVersion,
    result: ProjectSettingsResultV1,
}

impl ProjectSettingsResponseV1 {
    /// Creates the pathless no-project response.
    #[must_use]
    pub const fn no_project() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ProjectSettingsResultV1::NoProject,
        }
    }

    /// Creates a complete active-project response.
    #[must_use]
    pub fn available(settings: ActiveProjectSettingsV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: ProjectSettingsResultV1::Available {
                settings: Box::new(settings),
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::ConfirmProjectCommandAllowlistRequestV1;

    #[test]
    fn confirmation_request_rejects_core_owned_authority() {
        for forbidden in [
            "projectId",
            "worktreeId",
            "executable",
            "arguments",
            "confirmedAtUnixMillis",
        ] {
            let mut value = serde_json::json!({
                "protocolVersion": 1,
                "expectedCatalogId": "11".repeat(32),
                "expectedAllowlistRevision": null,
                "commandIds": ["22".repeat(32)]
            });
            value[forbidden] = serde_json::json!("forbidden");
            assert!(
                serde_json::from_value::<ConfirmProjectCommandAllowlistRequestV1>(value).is_err()
            );
        }
    }
}
